### PR TITLE
Allow excluding display drivers from automatic detection

### LINKED ISSUE #15196: Allow add-on braille display drivers to opt in for auto detection
### Is your feature request related to a problem? Please describe.
There is currently no documented way for braille display drivers to register themselves for auto detection. You can register a display in a global plugin, but as the global plugin loads after the braille subsystem, the display won't be detected in the first round of detection.

### Describe the solution you'd like
1. Add a `registerAutomaticDetection` method on `BrailleDisplayDriver`. This method must be implemented for drivers to register themselves in bdDetect. The method receives a new bdDetect.DriverRegistrar object that has addUsbDevices, addBluetoothDevices and addDeviceScanner methods that are bound to the driver.
2. Add a new checklistbox to the braille display selection dialog. When selection is set to automatic, this list box will be visible and allows you to opt out for auto detection of drivers you aren't interested in. For example, I only own HandyTech and Humanware devices, so the rest can be stolen from me.😉 Note that this list box will be a opt out list, so drivers will be enabled by default. This ensures that new drivers will always be enabled unless explicitly disabled.
3.  bdDetect.addUsbDevices and bdDetect.addBluetoothDevices will be deprecated. The advanced option to disable HID braille will also be removed in favor of this new ability to opt out for detection. As HID braille opt out is an advanced setting, I don't think profile upgrades are required.

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
closes #15196 

### Summary of the issue:
1. It is currently not possible to disable display drivers for automatic detection, other than the HID braille standard
2. There is no obvious way for add-on bundled braille display drivers to register themselves in automatic detection.

### Description of user facing changes
1. In the braille display selection dialog, when automatic is selected, there is now an additional checklistbox that allows you to disable drivers. The disabled drivers are saved in config.
2. The advanced option for to disable HID braille has been removed. Users should use the new checklistbox instead. There is a config profile upgrade step to remember the user's decision about explicitly disabling HID.
3. It is now possible to select the HID braille driver manually when it is disabled for automatic detection. The old option in advanced setting disabled HID completely, i.e. the driver wasn't available for manual selection either.

### Description of development approach
1. Added a registerAutomaticDetection classmethod to BrailleDisplayDriver. This method is called by bdDetect.initialize. The classmethod receives a bdDetect.DriverRegistrar that contains the following methods:
    * addUsbDevices: to register USB devices
    * addBluetoothDevices: To register bluetooth devices
    * addDeviceScanner: convenience method to register a handler to the scanForDevices extension point
2. Added a supportsAutomaticDetection boolean attribute on braille display drivers. This should be set to True for bdDetect to recognize the driver.
3. Removed the option from the config to enable/disable HID, instead rely on a new excludeDisplays lisst in the config that is populated by the new option in the braille display selection dialog. Note that exclusions are saved, not inclusions. This ensures that new drivers will be opt-out.
4. bdDetect.Detector._queueBgScan has a backwards compatible change for its limitToDevices parameter. When None, it falls back to the list of enabled drivers. When no drivers are disabled, it checks all drivers.
5. Logic from bdDetect.initialize() has been moved to the respective display drivers, calling the replacement methods on the driverRegistrar and omitting the name of the driver in the method call.

### Testing strategy:
Performed the following tests:

* Ensured that my APH mantis 40 is detected using both USB and Bluetooth
* Ensured that mantis wouldn't be detected when explicitly disabled
* Ensured that a config profile switch would detect a new display if an earlier detected display is disabled in a new profile, ensuring the disabled entry would be respected
* Ensured that HID braille is properly disabled as part of the config profile upgrade step

### Known issues with pull request:
None known

### Change log entries:
New features
- When using automatic detection of braille displays, it is now possible to opt-out drivers from detection from the Braille display Selection dialog. (#15196)
Changes
- The advanced setting to enable support for HID braille has been removed in favor of the new option to disable specific drivers for braille display auto detection in the braille display selection dialog. (#15196)
For Developers
- Added official support to register custom braille display drivers in the automatic braille display detection process. Consult the braille.BrailleDisplayDriver class documentation for more details. Most notably, the supportsAutomaticDetection attribute must be set to True and the registerAutomaticDetection classmethod must be implemented.  (#15196)
Deprecations
- bdDetect.addUsbDevices and bdDetect.addBluetoothDevices have been deprecated. Braille display drivers should implement the registerAutomaticDetection classmethod instead. That method receives a DriverRegistrar object on which the addUsbDevices and addBluetoothDevices methods can be used.
- The default implementation of the check method on BrailleDisplayDriver uses bdDetect.driverHasPossibleDevices for devices that are marked as thread safe. Starting from NVDA 2024.1, in order for the base method to use bdDetect.driverHasPossibleDevices, the supportsAutomaticDetection attribute must be set to True as well.

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

### CODE REVIEW

**Comment by leonardder** [comment]
It looks like #15208 also changed model names in the comments in bdDetect.py. I don't think that was a desirable change, therefore I left the dots in tact when fixing merge conflicts, effectively reverting that change.

**Comment by seanbudd** [comment]
@michaelDCurran - this and the issue #15196 still need a product decision

File: source/brailleDisplayDrivers/brailliantB.py [hunk 1]
```diff
@@ -84,6 +83,48 @@ class BrailleDisplayDriver(braille.BrailleDisplayDriver):
...
+			"APH Chameleon 20",
+			"APH Mantis Q40",
+			"Humanware BrailleOne",
+			"NLS eReader",
+			"NLS eReader Humanware",
+			"Brailliant BI 40X",
+			"Brailliant BI 20X",
+		)
+		)
+		)
```
Comment (seanbudd) [inline source/brailleDisplayDrivers/brailliantB.py:127]: Fix indentation for legiblity

```suggestion
		driverRegistrar.addBluetoothDevices(
			lambda m: (
				m.type == bdDetect.KEY_SERIAL
				and (
					m.id.startswith("Brailliant B")
					or m.id == "Brailliant 80"
					or "BrailleNote Touch" in m.id
				)
			)
			or (
				m.type == bdDetect.KEY_HID
				and m.deviceInfo.get("manufacturer") == "Humanware"
				and m.deviceInfo.get("product") in (
					"Brailliant HID",
					"APH Chameleon 20",
					"APH Mantis Q40",
					"Humanware BrailleOne",
					"NLS eReader",
					"NLS eReader Humanware",
					"Brailliant BI 40X",
					"Brailliant BI 20X",
				)
			)
		)
```

File: source/bdDetect.py [hunk 2]
```diff
@@ -99,48 +105,47 @@ def _isDebug():
...
+
+	def addBluetoothDevices(driver: str, matchFunc: MatchFuncT):
+		"""Associate Bluetooth HID or COM ports with a driver.
+		@param driver: The name of the driver.
+		@param matchFunc: A function which determines whether a given Bluetooth device matches.
+			It takes a L{DeviceMatch} as its only argument
+			and returns a C{bool} indicating whether it matched.
+		"""
+		log.warning(
+			"bdDetect.addBluetoothDevices is deprecated and wil lbe removed in NVDA 2024.1. "
```
Comment (seanbudd) [inline source/bdDetect.py:133]:
```suggestion
			"bdDetect.addBluetoothDevices is deprecated and will be removed in NVDA 2024.1. "
```

File: source/bdDetect.py [hunk 3]
```diff
@@ -545,267 +559,11 @@ def initialize():
...
-
-	# albatross
-	addUsbDevices("albatross", KEY_SERIAL, {
-		"VID_0403&PID_6001",  # Caiku Albatross 46/80
-	})
+	for display in getSupportedBrailleDisplayDrivers():
+		display.registerAutomaticDetection(DriverRegistrar(display.name))
+	# Hack, Caiku Albatross detection conflicts with other drivers
+	# when it isn't the last driver in the detection logic.
+	_driverDevices.move_to_end("albatross")
```
Comment (seanbudd) [inline source/bdDetect.py:566]: can this be removed now that #15269 has been merged?

**Review by seanbudd — COMMENTED** [review]
Thanks @leonardder for taking on this large project

File: source/brailleDisplayDrivers/brailliantB.py (hunk 1 above)
Comment (leonardder) [reply source/brailleDisplayDrivers/brailliantB.py:127]: I hope this keeps the lint happy. It was a drama to get this right!

File: source/brailleDisplayDrivers/hidBrailleStandard.py [hunk 4]
```diff
@@ -15,15 +15,7 @@
 import hwIo.hid
 from hwIo import intToByte
 
-from bdDetect import HID_USAGE_PAGE_BRAILLE
-
-
-def isSupportEnabled() -> bool:
```
[the remark this answers is not in the document]
Comment (leonardder) [reply source/brailleDisplayDrivers/hidBrailleStandard.py:21]: I restored the function. It calls bdDetect.driverIsEnabledForAutoDetection internally.

File: source/bdDetect.py (hunk 3 above)
Comment (leonardder) [reply source/bdDetect.py:566]: I'm afraid not. Nothing has changed in this area, i.e. placing albatross first will still mean a major slowdown for other devices as the driver takes so long to decide whether the display is valid.

File: source/bdDetect.py (hunk 2 above)
Comment (seanbudd) [reply source/bdDetect.py]: this hasn't been actioned

**Review by seanbudd — APPROVED** [review]

**Review by Qchristensen — APPROVED** [review]
User guide looks good, thanks!

**Comment by bramd** [comment]
The config migration seems to fail. When trying to update from a previous alpha build, I get:

```
Traceback (most recent call last):
  File "config\__init__.pyc", line 571, in _initBaseConf
  File "config\__init__.pyc", line 605, in _loadConfig
  File "config\__init__.pyc", line 601, in _loadConfig
  File "config\profileUpgrader.pyc", line 24, in upgrade
  File "config\profileUpgrader.pyc", line 41, in _doConfigUpgrade
  File "config\profileUpgradeSteps.pyc", line 368, in upgradeConfigFrom_10_to_11
  File "configobj\__init__.pyc", line 508, in __getitem__
KeyError: 'excludedDisplays'
```

**Pull request merged**